library/core/src/marker.rs

```diff
@@ -986,11 +986,16 @@ pub trait Tuple {}
 pub trait PointerLike {}
 
 /// A marker for types which can be used as types of `const` generic parameters.
+///
+/// These types must have a proper equivalence relation (`Eq`) and it must be automatically
```

Comment:
Is this "just in case"? Like, there are no types that implement `StructuralEq` that don't also implement `Eq`, right, at least without a manual `StructuralEq` impl?

Reply:
Would be nice if you could add a test showing where possibly dubious code goes from pass -> err.

Reply:
A generic type with `derive(Eq)` will implement `StructuralEq` without any bounds on the generic parameters.

This is mostly to make the 2nd half of the comment coherent: we want to say that valtree equality agrees with `==`. Without at least a `PartialEq` bound, that statement is ill-formed, since the type might not even have `==`.

Reply:
> Would be nice if you could add a test showing where possibly dubious code goes from pass -> err.


I don't know any such test, but currently the invariants that achieve that `ConstParamTy` implies `Eq` are rather subtle. This makes them a lot more obvious.

Reply:
> currently the invariants that achieve that ConstParamTy implies Eq are rather subtle. This makes them a lot more obvious.

Yeah, ok, that's all I wanted to really know. This isn't necessarily for "correctness" (but perhaps only by accident do we imply the right thing due to recursively checking fields for `ConstParamTy` in impls).

Reply:
I also view this as preparation for [removing the `StructuralEq` trait entirely](https://github.com/rust-lang/rust/issues/115881) (at which point only the `StructuralPartialEq` bound would be left otherwise).

I can do that all in the same PR if you prefer.

Reply:
Nah, I'll probably have more questions about behavioral changes on that PR anyways 😅